File: ios/SsdidWallet/Domain/OpenId4Vp/MDocVpTokenBuilder.swift
import Foundation

/// Builds a CBOR DeviceResponse for mdoc presentation per ISO 18013-5.
///
/// DeviceResponse = {
///   "version": "1.0",
///   "documents": [{ docType, issuerSigned, deviceSigned }],
///   "status": 0
/// }
enum MDocVpTokenBuilder {

    /// Builds a VP token from a stored mdoc with selective disclosure.
    ///
    /// - Parameters:
    ///   - storedMDoc: The stored mdoc credential
    ///   - requestedElements: Map of namespace to requested element identifiers
    ///   - clientId: The verifier's client_id
    ///   - responseUri: The verifier's response_uri
    ///   - nonce: Verifier-provided nonce
    ///   - signer: Function that signs DeviceAuthentication bytes
    /// - Returns: Base64url-encoded CBOR DeviceResponse string
    static func build(
        storedMDoc: StoredMDoc,
        requestedElements: [String: [String]],
        clientId: String,
        responseUri: String,
        nonce: String,
        signer: (Data) -> Data
    ) -> String? {
        // Parse the stored IssuerSigned CBOR
        guard let issuerSigned = MsoParser.parseIssuerSigned(storedMDoc.issuerSignedCbor) else {
            return nil
        }

        // Apply selective disclosure
        let presented = MDocPresenter.present(
            issuerSigned: issuerSigned,
            requestedElements: requestedElements
        )

        // Build SessionTranscript
        let sessionTranscript = SessionTranscript.build(
            clientId: clientId,
            responseUri: responseUri,
            nonce: nonce
        )

        // Build DeviceAuthentication = ["DeviceAuthentication", SessionTranscript, docType, {}]
        let deviceAuth: [Any] = [
            "DeviceAuthentication",
            sessionTranscript, // raw CBOR bytes
            storedMDoc.docType,
            [String: Any]() // empty DeviceNameSpaces
        ]
        let deviceAuthBytes = CborCodec.encodeDataElement(deviceAuth)

        // Sign DeviceAuthentication
        let signature = signer(deviceAuthBytes)

        // Build COSE_Sign1 for deviceSignature: [protectedHeaders, unprotectedHeaders, null, signature]
        // Encode as CBOR array with tag 18
        let coseSign1Bytes = encodeCoseSign1(signature: signature)

        // Build DeviceSigned map
        let emptyNameSpaces = CborCodec.encodeMap([:])
        let deviceSignedMap: [String: Any] = [
            "nameSpaces": emptyNameSpaces, // tagged bstr
            "deviceAuth": ["deviceSignature": coseSign1Bytes] as [String: Any]
        ]

        // Build IssuerSigned CBOR for response
        let issuerSignedCbor = encodeIssuerSigned(presented)

        // Build Document
        let document: [String: Any] = [
            "docType": storedMDoc.docType,
            "issuerSigned": issuerSignedCbor,
            "deviceSigned": deviceSignedMap
        ]

        // Build DeviceResponse
        let deviceResponse: [String: Any] = [
            "version": "1.0",
            "documents": [document],
            "status": 0
        ]

        let responseBytes = CborCodec.encodeMap(deviceResponse)
        return base64urlEncode(responseBytes)
    }

    // MARK: - Private

    private static func encodeCoseSign1(signature: Data) -> Data {
        // COSE_Sign1 = [bstr, {}, null, bstr] as a CBOR array
        var buffer = Data()
        // Array of 4 items
        buffer.append(0x84)
        // protected: empty bstr
        buffer.append(0x40)
        // unprotected: empty map
        buffer.append(0xa0)
        // payload: null (detached)
        buffer.append(0xf6)
        // signature: bstr
        let sigBytes = CborCodec.encodeDataElement(signature)
        buffer.append(sigBytes)
        return buffer
    }

    private static func encodeIssuerSigned(_ issuerSigned: IssuerSigned) -> Data {
        var nameSpacesDict = [String: Any]()
        for (ns, items) in issuerSigned.nameSpaces {
            var encodedItems = [Any]()
            for item in items {
                let itemMap: [String: Any] = [
                    "digestID": item.digestId,
                    "random": item.random,
                    "elementIdentifier": item.elementIdentifier,
                    "elementValue": item.elementValue
                ]
                // Each item is tagged bstr (tag 24) containing the encoded item
                let itemBytes = CborCodec.encodeMap(itemMap)
                encodedItems.append(itemBytes)
            }
            nameSpacesDict[ns] = encodedItems
        }

        let result: [String: Any] = [
            "nameSpaces": nameSpacesDict,
            "issuerAuth": issuerSigned.issuerAuth
        ]
        return CborCodec.encodeMap(result)
    }

    private static func base64urlEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .trimmingCharacters(in: CharacterSet(charactersIn: "="))
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/oid4vp/MDocVpTokenBuilder.kt
package my.ssdid.wallet.domain.oid4vp

import com.upokecenter.cbor.CBORObject
import my.ssdid.wallet.domain.mdoc.CborCodec
import my.ssdid.wallet.domain.mdoc.IssuerSigned
import my.ssdid.wallet.domain.mdoc.MDocPresenter
import my.ssdid.wallet.domain.mdoc.MsoParser
import my.ssdid.wallet.domain.mdoc.StoredMDoc

/**
 * Builds a CBOR DeviceResponse for mdoc presentation per ISO 18013-5.
 *
 * DeviceResponse = {
 *   "version": "1.0",
 *   "documents": [{ docType, issuerSigned, deviceSigned }],
 *   "status": 0
 * }
 */
object MDocVpTokenBuilder {

    fun build(
        storedMDoc: StoredMDoc,
        requestedElements: Map<String, List<String>>,
        clientId: String,
        responseUri: String,
        nonce: String,
        signer: (ByteArray) -> ByteArray
    ): String {
        // Parse the stored IssuerSigned CBOR
        val issuerSigned = MsoParser.parseIssuerSigned(storedMDoc.issuerSignedCbor)

        // Apply selective disclosure
        val presented = MDocPresenter.present(issuerSigned, requestedElements)

        // Build SessionTranscript
        val sessionTranscript = SessionTranscript.build(clientId, responseUri, nonce)

        // Build DeviceAuthentication = ["DeviceAuthentication", SessionTranscript, docType, {}]
        val deviceAuth = CBORObject.NewArray()
        deviceAuth.Add(CBORObject.FromObject("DeviceAuthentication"))
        deviceAuth.Add(CBORObject.DecodeFromBytes(sessionTranscript))
        deviceAuth.Add(CBORObject.FromObject(storedMDoc.docType))
        deviceAuth.Add(CBORObject.NewMap()) // empty DeviceNameSpaces

        // Wrap DeviceAuthentication in tag 24 (encoded CBOR data item) per ISO 18013-5 §9.1.3
        val deviceAuthTagged = CBORObject.FromObjectAndTag(deviceAuth.EncodeToBytes(), 24)

        // Build COSE Sig_structure per RFC 9052 §4.4 and ISO 18013-5 §9.1.3:
        // Sig_structure = ["Signature1", body_protected, external_aad, payload]
        // - external_aad is empty (SessionTranscript is already in DeviceAuthentication)
        // - payload is the tag-24-wrapped DeviceAuthentication bstr
        val protectedHeaders = byteArrayOf() // empty protected headers
        val sigStructure = CBORObject.NewArray()
        sigStructure.Add(CBORObject.FromObject("Signature1"))
        sigStructure.Add(CBORObject.FromObject(protectedHeaders))
        sigStructure.Add(CBORObject.FromObject(byteArrayOf())) // external_aad: empty
        sigStructure.Add(CBORObject.FromObject(deviceAuthTagged.EncodeToBytes())) // payload: #6.24(DeviceAuthentication)

        // Sign the Sig_structure
        val signature = signer(sigStructure.EncodeToBytes())

        // Build COSE_Sign1 for deviceSignature
        val coseSign1 = CBORObject.NewArray()
        coseSign1.Add(CBORObject.FromObject(protectedHeaders)) // protected
        coseSign1.Add(CBORObject.NewMap()) // unprotected
        coseSign1.Add(CBORObject.Null) // payload is detached
        coseSign1.Add(CBORObject.FromObject(signature))

        // Build DeviceSigned
        val deviceSigned = CBORObject.NewMap()
        deviceSigned["nameSpaces"] = CBORObject.FromObjectAndTag(
            CBORObject.NewMap().EncodeToBytes(), 24
        )
        val deviceAuthMap = CBORObject.NewMap()
        deviceAuthMap["deviceSignature"] = CBORObject.FromObjectAndTag(coseSign1, 18)
        deviceSigned["deviceAuth"] = deviceAuthMap

        // Build the IssuerSigned CBOR for the response
        val issuerSignedCbor = encodeIssuerSigned(presented)

        // Build Document
        val document = CBORObject.NewMap()
        document["docType"] = CBORObject.FromObject(storedMDoc.docType)
        document["issuerSigned"] = issuerSignedCbor
        document["deviceSigned"] = deviceSigned

        // Build DeviceResponse
        val deviceResponse = CBORObject.NewMap()
        deviceResponse["version"] = CBORObject.FromObject("1.0")
        val documents = CBORObject.NewArray()
        documents.Add(document)
        deviceResponse["documents"] = documents
        deviceResponse["status"] = CBORObject.FromObject(0)

        // Base64url encode the CBOR bytes
        val responseBytes = deviceResponse.EncodeToBytes()
        return java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(responseBytes)
    }

    private fun encodeIssuerSigned(issuerSigned: IssuerSigned): CBORObject {
        val cbor = CBORObject.NewMap()
        val nameSpaces = CBORObject.NewMap()
        for ((ns, items) in issuerSigned.nameSpaces) {
            val array = CBORObject.NewArray()
            for (item in items) {
                val itemCbor = CBORObject.NewMap()
                itemCbor["digestID"] = CBORObject.FromObject(item.digestId)
                itemCbor["random"] = CBORObject.FromObject(item.random)
                itemCbor["elementIdentifier"] = CBORObject.FromObject(item.elementIdentifier)
                itemCbor["elementValue"] = CborCodec.toCborObject(item.elementValue)
                array.Add(CBORObject.FromObjectAndTag(itemCbor.EncodeToBytes(), 24))
            }
            nameSpaces[ns] = array
        }
        cbor["nameSpaces"] = nameSpaces
        cbor["issuerAuth"] = CBORObject.DecodeFromBytes(issuerSigned.issuerAuth)
        return cbor
    }
}
